### DISCUSSION TITLE
Need more clarity on the intended use case

### DISCUSSION THREAD

**Discussion by dvnlgls** [pocketbase/pocketbase#6079] [Q&A]
I spent an entire evening playing with Pocketbase and was quite impressed with the simplicity of the stack. But I'm a little confused with the exact use case despite reading past answers. 

For instance, in the post https://github.com/pocketbase/pocketbase/discussions/5479 @ganigeorgiev explains it as:

> while PocketBase can be used as a backend for websites at the moment we don't have the necessary helpers for that and depending on your use case it may not be the best tool for the job. PocketBase was designed primarily for building applications that usually can be interact directly from the client-side (e.g. SPA in the browser or a mobile app).

The part about "backend for website" is clear. I use Node for all my backend so I get that. But does "PocketBase was designed primarily for building applications that usually can be interact directly from the client-side" mean clients directly interacting with the **_DB_** Pocketbase provides? (along with other built-in features like auth, file upload etc)

-------

That's the main question I'm trying to sort out. Since I asked for more clarity, I'll explain my use cases and understanding:

My intended use case:
- Mostly damn simple side projects and proof of concepts (All CRUD apps. Only slightly complex than to-do apps.) A simple relational db to to do only the most basic CRUD operations. No need for SSR or any fancy things.
- My go to stack is: some front-end like Angular of pure JS, a Node/express backend with MySQL. 

Why am I considering Pocketbase:
- Simplicity. I'm quite comfortable with my stack but setting a backend with a DB like MySQL seems an overkill for most of my projects. Despite having templates and a detailed knowledge base, it's still a lot of work setting everything up.
- Sqlite is perfect for many of my projects.
- _I really like the simplicity of just dropping an executable with some basic file to handle a few API calls on some cheap VPS, run a `serve` command with a process monitor and call it a day._ 

My understanding of Pocketbase:
- I think there are two ways to use Pocketbase: as a framework (using JS hooks) or as a client side sdk (as a wrapper around the remotely hosted Pocketbase sqlite DB and other features. I think the SDK can also be used on a backend Node server?)

- I gather the framework method has some limitations because it's not using a V8 engine and some form of a Go port of JS?
    - Yet to read all the limitations but for just getting basic API requests from the client, querying the db and sending back a response (setting http headers and such), _is this method sufficient or will I run into major roadblocks_?
    - Will I be able to use npm packages in this approach?
    - _If I'm going this route, would I be better off just writing a simple Node/Express app to manage a SQLite and API calls?_
    
- If I go the sdk route, then I'd need Pocketbase (hosted on some VPS) just to manage the db. No need to write any server logic like handling APIs, since they are dealt from the client.
    - **The part that confuses me the most about the SDK approach** is: if I'm running pocketbase on a server, why can't I just use Node and deal with SQLite directly? I'm sure I'm missing a justifiable use case; hence my post. (does the approach make more sense in GO parlance of which I have no idea?)

Is my understanding accurate?

To rephrase my question: compared to a traditional client-server (in JS) approach what precise time-saving or quality of life improvement features that Pocketbase offers? (I'm absolutely not crapping on the project but trying to understand the project's strengths to make decisions.)

--------
If this topic has been covered already at depth answering my questions, please point me to the right place. Else, some clarity would be greatly appreciated.

**Comment by ganigeorgiev** [comment]
Please spend some time reading the documentation and more specifically:
- https://pocketbase.io/faq/
- https://pocketbase.io/docs/client-side-sdks/
- https://pocketbase.io/docs/use-as-framework/

If after reading the documentation you are still not sure how to use PocketBase and what is it intended for, then I'm out of ideas how to explain it better.

**This is NOT a commercial project, no one is trying to sell you anything or "convince" you to use it.**

**Reply by dvnlgls** [reply]
@ganigeorgiev I was very clear that I'm not crapping on the project and that I took time to not only read past answers but also write a detailed thought process and consideration. I NEVER wanted any convincing (not sure why you got that impression.)

I took the time to post this question and not a simple "send halp" because I believe my questions are valid. Your non-answer, bordering on insult, is counter productive and I can't see how it can possibly help your project and prospective users. I suggest you reconsider your attitude. Not everyone has your specific frame of thinking and if you can't be bothered to answer at least let others deal with questions that may be beneath you. 

P.S.: If you genuinely feel you are out of ideas on documentation, say it so and ask for help. No documentation can cover all questions, that's why we have support channels. No need to be rude to people eager and trying to use your software.

**Reply by ganigeorgiev** [reply]
Fine. I'll reopen it.

**Reply by saikoml** [reply]
**** you @ganigeorgiev

**Reply by ganigeorgiev** [reply]
@saikoml I don't know what have triggered the above reply, but this type of language is definitely not welcomed. Next time you'll be banned.

If it was a reaction to me being "rude" then it is exaggerated.

I didn't answer the questions from the initial post because all of them are too vague to have a definite answer and will require several back-and-forth to get to the actual point and I no longer have energy for this.

I simply think that the big red WHY section in the FAQ is already clear enough and I find this type of discussion pointless.

The author is asking for a "justifiable use case", hence why it reads to me as they are looking for "convincing" why to use PocketBase.

In general, the shortest and safest answer would be to not use PocketBase if you don't find it useful or not sure why do you need it in the first place.

In any case, there is no need for profanity or arguing under my reply. If you think that you can help @dvnlgls then please answer him.

If you or someone else think that something in the documentation could be clarified better, then feel free to open a PR with your suggestion in the [docs repo](https://github.com/pocketbase/site) and I'll try to review it.

**Reply by dvnlgls** [reply]
I agree that profanity is uncalled for and it's also unhelpful. No place for that in civil discourse.

I disagree that "all my questions are too vague". Based on the documentation, I explained my understanding and the considerations for taking one of the two approaches the project offers (hooks and sdk.) 

> In general, the shortest and safest answer would be to not use PocketBase if you don't find it useful or not sure why do you need it in the first place.

I'm struggling to understand this take. How can one make a decision if the fundamental questions aren't cleared? For reference, before coming here, I tried supabase and cloudflare workers, and decided they don't work for me. Then I read about pocketbase on reddit and thought it was interesting to consider as a serious candidate.

If this project is something you built for your specific use case and releasing it to the public with a "it's what it is. take it or leave it" intention, then I totally get it. I have a few such projects of my own. But please make it clear. That would save both you and curious minds like me some time and effort. 

I'm inclined to believe you didn't even read my post but that's your prerogative. While I quite enjoyed playing with your idea, I no longer wish to spend any more time when the founder himself doesn't want me to. Surprisingly even on StackOverflow I haven't been dismissed for asking a question. Oh well! 

(If someone does answer, maybe future users may find it helpful.)

**Comment by pnmcosta** [comment]
My 5 cents on this discussion with my experience after using PB for the past year or so building a CRUD Shopify App that has been live since Jan and is now in production with the latest v0.23.

My use case is similar to the one mentioned by the OP, no SSR needed just some endpoints that my React app calls, additionally internal processing, and background tasks to consume external message queues.

Some of the reasons I picked PB are enumerated below (in no particular order)

- OTB auth providers
- Performant Sqlite
- DB schema management and migrations
- ORM like features, e.g. [this](https://github.com/pocketbase/pocketbase/discussions/5702)
- CRON jobs
- Ability to plugin whatever GO code/libraries into it (my app for example consumes AWS SQS queues within the same running process)
- Server sent events to rehydrate React app state when those change.

It was a no brainer for me having this open source framework that I could get to understand the internals of, and leverage for my needs within its constraints - a monolith architecture.

It is my opinion that if you're from a GO background it would be much easier to understand its purpose and how/why it works the way it does, I have many times dug through the code to understand what is happening and how best to work with/around it.

I personally recommend it to anyone that is starting a new project because the overhead costs of running it are minimal in comparison to other BaaS out there.

I honestly do not understand why JS hooks even exist in the first place 😅 I think this just causes a lot of confusion, but this is just my opinion, as a GO developer I have no use for them. However I'm sure there's a use case for them that I'm not considering at all.

@ganigeorgiev has his personality and takes, but I would not have been able to achieve as quickly as a I did without PB's help, so I understand and respect where he's coming from. 

In sum, and this is an over simplification, PB is a "DB over REST" framework that you can tweak and extend for your needs with easy deployment and cheap running costs.

**Reply by suelitonx** [reply]
Many people don't use or don't like Go, others have been using JavaScript their whole lives and a "js hook" would be more convenient. This is what we call accessibility and inclusion, lol.

**Reply by pnmcosta** [reply]
Fair enough, if there's a need for it and is feasible (as it's been proven to be) I can understand and respect that. 

However, I do have an issue with this:

> Many people don't use or don't like Go, others have been using JavaScript their whole lives and a "js hook" would be more convenient.

This is an open source project built with Go, I think that if anyone is considering using it to its full potential they should at least have a minimal understanding of its internals before jumping into it. 

The project has great documentation (some of the best I've seen) but certain edge cases you need to jump into the code, debug and troubleshoot it, without minimal go understanding it would be tricky to move forward, at least I think I would 😅

**Comment by dvnlgls** [comment]
@pnmcosta Thank you for taking the time to provide an actual helpful answer. I suspected Go folks are the primary target for PB and from what you said, that seems to be the case. 

I stumbled on PB while looking for a JS solution and naturally, hooks is the first thing I tried and then started wondering if the limitations were big enough to prevent me from running a simple API server, which led to a rabbit hole, some confusion and finally leading to this post. 

I totally respect that things make much sense in Go. Not everything has to be JS-centric. After all, a JS framework/solution is one thing the world is short of ;)

**Reply by pnmcosta** [reply]
You're welcome. I really hope I didn't come across as putting a divide between JS and GO folk, if that's the understanding from my answer, I apologize - it was unintentional - I have personally coded with JS and .NET before, but I really feel in love with Go's syntax, simplicity and boilerplate-less nature of the language, I feel a lot more in control of what's happening rather than hope that it works correctly (but this is too an oversimplification! 😅). 

I mentioned JS hooks (which I regret now) purely because it's clear it causes some confusion by scanning the discussion board and only after you get to grips with the framework itself, then it clicks. But be aware that hooks are totally optional, you would only need to leverage hooks if you want to customize something between a request and a DB CRUD op. 

IMO what you describe (a simple API server) with clearly defined data models I really doubt you would need to use hooks. I'm 99.9% sure that the majority of todo-app-like-examples would not need to leverage hooks at all.

If PB is definitely not for you, there is one project that I've worked with before a while back and had a solid experience with https://www.totaljs.com/ for a more JS centric ecosystem.

Best of luck, and if you do opt for PB instead you'll be more than welcome back here.

**Reply by dvnlgls** [reply]
@pnmcosta Hey, you did NOT put a divide between JS and GO! At all! Your intentions, message and perspectives were crystal clear, and I found them helpful. If anything, you've given me the idea to consider GO for web apps. JS is so synonymous with web that it didn't even occur to me to consider alternatives. 

I think I'm getting the picture. So far I've never had a need for such a client SDK (or ORM), hence the confusion. I prefer writing direct SQL queries on the backend and keeping the communication between the client and server as simple as possible through a few API calls. (that is, for the hobby projects at least.) So, I think I may not benefit from PB.

P.S: Thanks for providing an alternative, totaljs. I think we need a search engine just to index JS stack out there! And also, for understanding people's considerations and exhibiting a welcoming nature.
